### PR TITLE
ugrade connection pool

### PR DESCRIPTION
**Pull request by KlotzAndrew**
issues we're solving:

* gc/reaper class was killing in-use connections, causing dropped messages

cannot close connection in `trap_singals!` method due to`ThreadError` trying to access another
thread outside the trap handler

### CODE REVIEW

**Comment by sachinsiby** [comment]
@maedhr Do we know why that reaper was written in the first place?

File: lib/multiple_man/runner.rb [hunk 1]
```diff
@@ -28,9 +27,6 @@ def run
     def trap_signals!
       handler = proc do |signal|
         puts "received #{Signal.signame(signal)}"
-        channel.close
-        connection.close
-
```
Comment (whoward) [inline lib/multiple_man/runner.rb]: if we're dropping the channel and connection closing we might as well drop `trap_signals!` entirely - it's just printing a message at this point

File: lib/multiple_man/model_publisher.rb [hunk 2]
```diff
@@ -24,17 +23,22 @@ def publish(records, operation=:create)
...
 
       MultipleMan.logger.debug("  Record Data: #{data} | Routing Key: #{routing_key}")
 
-      connection.topic.publish(data.payload, routing_key: routing_key)
+      publish_payload(channel, data.payload, routing_key)
+    end
+
+    def publish_payload(channel, payload, routing_key)
+      topic = channel.topic(MultipleMan.configuration.topic_name, MultipleMan.configuration.exchange_opts)
+      topic.publish(payload, routing_key: routing_key)
```
Comment (whoward) [inline lib/multiple_man/model_publisher.rb]: just a general question - why the separate method here? it's only called in one place...I don't think rubocop would have fought you on keeping it in one method

File: lib/multiple_man/connection.rb [hunk 3]
```diff
@@ -1,109 +1,55 @@
...
-        Thread.current[:multiple_man_exception_retry_count] = 0
-        raise ConnectionError, e
+    class << self
+      def channel_pool
+        @channel_pool ||= ConnectionPool.new(size: 5) do
+          channel = bunny_connection.create_channel
+          channel.confirm_select if MultipleMan.configuration.publisher_confirms
+          channel.topic(topic_name, MultipleMan.configuration.exchange_opts)
+          channel
+        end
```
Comment (whoward) [inline lib/multiple_man/connection.rb]: i'm not expert with ruby threading but i think this needs mutual exclusion

File: lib/multiple_man/connection.rb [hunk 4]
```diff
@@ -1,109 +1,55 @@
...
-      retry_count += 1
-
-      if retry_count < recovery_options[:max_retries]
-        Thread.current[:multiple_man_exception_retry_count] = retry_count
-        sleep recovery_options[:time_between_retries]
-        retry
-      else
-        Thread.current[:multiple_man_exception_retry_count] = 0
-        raise ConnectionError, e
+    class << self
```
Comment (whoward) [inline lib/multiple_man/connection.rb]: you can probably replace this with `module_function` to save yourself an indent

File: lib/multiple_man/connection.rb [hunk 5]
```diff
@@ -1,109 +1,55 @@
...
+      def channel_pool
+        @channel_pool ||= ConnectionPool.new(size: 5) do
```
Comment (whoward) [inline lib/multiple_man/connection.rb]: connection pool size should be configurable

File: lib/multiple_man/runner.rb (hunk 1 above)
Comment (sachinsiby) [reply lib/multiple_man/runner.rb]: Shouldn't `MultipleMan::Connection.close` be called here?

File: lib/multiple_man/connection.rb [hunk 6]
```diff
@@ -1,109 +1,55 @@
...
-
-    def self.reset!
-      @mutex.synchronize do
-        @connection.close if @connection
-        @connection = nil
 
-        @channel_gc.stop if @channel_gc
-        @channel_gc = nil
+      def mutex
+        @mutex = Mutex.new
```
Comment (bradrobertson) [inline lib/multiple_man/connection.rb]: I don't think this is doing what you want. Every call to `mutex` yields a new object, that kind of defeats the purpose of a lock.

File: lib/multiple_man/connection.rb (hunk 5 above)
Comment (bradrobertson) [reply lib/multiple_man/connection.rb]: `||=` is not threadsafe. Is there a reason we want to limit the pool of channels? My understanding from AMQP is that these are lightweight and can just be created/destroyed at will. I don't think channels are ever meant to be reused

Comment (bradrobertson) [reply lib/multiple_man/connection.rb]: I suppose the previous impl also re-used the channels for each thread but again, perhaps we should look into that in depth to see if that's really how it should work

File: lib/multiple_man/runner.rb (hunk 1 above)
Comment (whoward) [reply lib/multiple_man/runner.rb]: yeah and probably shut down the pool too

File: lib/multiple_man/connection.rb (hunk 5 above)
Comment (bradrobertson) [reply lib/multiple_man/connection.rb]: Ok ignore this. According to [this](http://rubybunny.info/articles/connecting.html#opening_a_channel) channels _should_ be long lived

File: lib/multiple_man/model_publisher.rb (hunk 2 above)
Comment (bradrobertson) [reply lib/multiple_man/model_publisher.rb]: is there any concern declaring the exchange on each push? I'm assuming this has some call to Rabbit itself. It'd be nice to have some benchmarks around this to understand if there's a hit in any way.

File: lib/multiple_man/runner.rb (hunk 1 above)
Comment (bradrobertson) [reply lib/multiple_man/runner.rb]: Agreed... I think `close` does shut the pool down though.

**Comment by bradrobertson** [comment]
@KlotzAndrew It would be worth going through this history. Interestingly this has been done before (though not necessarily properly)
```
± br |master ✓| → git log -S channel_pool
commit ad1f3e77425c3fe3521fdf1e209b2c95d0bb6d41
Author: Craig Savolainen and Zulfiqar Ali <dev+craig+zulfiqar@influitive.com>
Date:   Tue May 26 14:08:21 2015 -0400

    Operations on channel while session down may lose operations.

    Stashing the channel in a thread local variable removes the ability
    to know when it is no longer needed. MRI GC may not reap the object
    in quickly enough so a secondary GC has been added to watch for
    thread death and close channels.

commit 085acfc7c03e00e0ba70f7cf1e9ae466cf493b3d
Author: Ryan Brunner <ryan@influitive.com>
Date:   Fri Sep 5 07:44:08 2014 -0400

    Use thread.current for channels

commit b0d3e10d9270f4c03f8fdfaa6cd80d4e9c5697b0
Author: Ryan Brunner <ryan@influitive.com>
Date:   Mon Apr 7 16:44:08 2014 -0400

    Trying to get one connection with concurrency

commit 020262e1aa7fb314da787f34133eac2fcee8fd53 (tag: v0.5.3)
Author: Ryan Brunner <ryan@influitive.com>
Date:   Mon Apr 7 16:33:13 2014 -0400

    Adding worker concurrency to listeners

commit 02eb95e90c772b7858253fe1869065bf098faaea (tag: v0.2.7)
Author: Ryan Brunner <ryan@influitive.com>
Date:   Thu Jan 9 06:33:30 2014 -0500

    Added 'as' capability, channel pool and error logging
```

**Pull request closed without merging**